// @google-cloud/documentai/build/src/v1beta3/document_service_client.js
"use strict";
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentServiceClient = void 0;
const jsonProtos = require("../../protos/protos.json");
/**
 * Client JSON configuration object, loaded from
 * `src/v1beta3/document_service_client_config.json`.
 * This file defines retry strategy and timeouts for all API methods in this library.
 */
const gapicConfig = require("./document_service_client_config.json");
const version = require('../../../package.json').version;
/**
 *  Service to call Cloud DocumentAI to manage document collection (dataset).
 * @class
 * @memberof v1beta3
 */
class DocumentServiceClient {
    /**
     * Construct an instance of DocumentServiceClient.
     *
     * @param {object} [options] - The configuration object.
     * The options accepted by the constructor are described in detail
     * in [this document](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#creating-the-client-instance).
     * The common options are:
     * @param {object} [options.credentials] - Credentials object.
     * @param {string} [options.credentials.client_email]
     * @param {string} [options.credentials.private_key]
     * @param {string} [options.email] - Account email address. Required when
     *     using a .pem or .p12 keyFilename.
     * @param {string} [options.keyFilename] - Full path to the a .json, .pem, or
     *     .p12 key downloaded from the Google Developers Console. If you provide
     *     a path to a JSON file, the projectId option below is not necessary.
     *     NOTE: .pem and .p12 require you to specify options.email as well.
     * @param {number} [options.port] - The port on which to connect to
     *     the remote host.
     * @param {string} [options.projectId] - The project ID from the Google
     *     Developer's Console, e.g. 'grape-spaceship-123'. We will also check
     *     the environment variable GCLOUD_PROJECT for your project ID. If your
     *     app is running in an environment which supports
     *     {@link https://developers.google.com/identity/protocols/application-default-credentials Application Default Credentials},
     *     your project ID will be detected automatically.
     * @param {string} [options.apiEndpoint] - The domain name of the
     *     API remote host.
     * @param {gax.ClientConfig} [options.clientConfig] - Client configuration override.
     *     Follows the structure of {@link gapicConfig}.
     * @param {boolean} [options.fallback] - Use HTTP/1.1 REST mode.
     *     For more information, please check the
     *     {@link https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#http11-rest-api-mode documentation}.
     * @param {gax} [gaxInstance]: loaded instance of `google-gax`. Useful if you
     *     need to avoid loading the default gRPC version and want to use the fallback
     *     HTTP implementation. Load only fallback version and pass it to the constructor:
     *     ```
     *     const gax = require('google-gax/build/src/fallback'); // avoids loading google-gax with gRPC
     *     const client = new DocumentServiceClient({fallback: true}, gax);
     *     ```
     */
    constructor(opts, gaxInstance) {
        var _a, _b, _c, _d, _e;
        this._terminated = false;
        this.descriptors = {
            page: {},
            stream: {},
            longrunning: {},
            batching: {},
        };
        // Ensure that options include all the required fields.
        const staticMembers = this.constructor;
        if ((opts === null || opts === void 0 ? void 0 : opts.universe_domain) &&
            (opts === null || opts === void 0 ? void 0 : opts.universeDomain) &&
            (opts === null || opts === void 0 ? void 0 : opts.universe_domain) !== (opts === null || opts === void 0 ? void 0 : opts.universeDomain)) {
            throw new Error('Please set either universe_domain or universeDomain, but not both.');
        }
        const universeDomainEnvVar = typeof process === 'object' && typeof process.env === 'object'
            ? process.env['GOOGLE_CLOUD_UNIVERSE_DOMAIN']
            : undefined;
        this._universeDomain =
            (_c = (_b = (_a = opts === null || opts === void 0 ? void 0 : opts.universeDomain) !== null && _a !== void 0 ? _a : opts === null || opts === void 0 ? void 0 : opts.universe_domain) !== null && _b !== void 0 ? _b : universeDomainEnvVar) !== null && _c !== void 0 ? _c : 'googleapis.com';
        this._servicePath = 'documentai.' + this._universeDomain;
        const servicePath = (opts === null || opts === void 0 ? void 0 : opts.servicePath) || (opts === null || opts === void 0 ? void 0 : opts.apiEndpoint) || this._servicePath;
        this._providedCustomServicePath = !!((opts === null || opts === void 0 ? void 0 : opts.servicePath) || (opts === null || opts === void 0 ? void 0 : opts.apiEndpoint));
        const port = (opts === null || opts === void 0 ? void 0 : opts.port) || staticMembers.port;
        const clientConfig = (_d = opts === null || opts === void 0 ? void 0 : opts.clientConfig) !== null && _d !== void 0 ? _d : {};
        const fallback = (_e = opts === null || opts === void 0 ? void 0 : opts.fallback) !== null && _e !== void 0 ? _e : (typeof window !== 'undefined' && typeof (window === null || window === void 0 ? void 0 : window.fetch) === 'function');
        opts = Object.assign({ servicePath, port, clientConfig, fallback }, opts);
        // Request numeric enum values if REST transport is used.
        opts.numericEnums = true;
        // If scopes are unset in options and we're connecting to a non-default endpoint, set scopes just in case.
        if (servicePath !== this._servicePath && !('scopes' in opts)) {
            opts['scopes'] = staticMembers.scopes;
        }
        // Load google-gax module synchronously if needed
        if (!gaxInstance) {
            gaxInstance = require('google-gax');
        }
        // Choose either gRPC or proto-over-HTTP implementation of google-gax.
        this._gaxModule = opts.fallback ? gaxInstance.fallback : gaxInstance;
        // Create a `gaxGrpc` object, with any grpc-specific options sent to the client.
        this._gaxGrpc = new this._gaxModule.GrpcClient(opts);
        // Save options to use in initialize() method.
        this._opts = opts;
        // Save the auth object to the client, for use by other methods.
        this.auth = this._gaxGrpc.auth;
        // Set useJWTAccessWithScope on the auth object.
        this.auth.useJWTAccessWithScope = true;
        // Set defaultServicePath on the auth object.
        this.auth.defaultServicePath = this._servicePath;
        // Set the default scopes in auth client if needed.
        if (servicePath === this._servicePath) {
            this.auth.defaultScopes = staticMembers.scopes;
        }
        this.locationsClient = new this._gaxModule.LocationsClient(this._gaxGrpc, opts);
        // Determine the client header string.
        const clientHeader = [`gax/${this._gaxModule.version}`, `gapic/${version}`];
        if (typeof process === 'object' && 'versions' in process) {
            clientHeader.push(`gl-node/${process.versions.node}`);
        }
        else {
            clientHeader.push(`gl-web/${this._gaxModule.version}`);
        }
        if (!opts.fallback) {
            clientHeader.push(`grpc/${this._gaxGrpc.grpcVersion}`);
        }
        else {
            clientHeader.push(`rest/${this._gaxGrpc.grpcVersion}`);
        }
        if (opts.libName && opts.libVersion) {
            clientHeader.push(`${opts.libName}/${opts.libVersion}`);
        }
        // Load the applicable protos.
        this._protos = this._gaxGrpc.loadProtoJSON(jsonProtos);
        // This API contains "path templates"; forward-slash-separated
        // identifiers to uniquely identify resources within the API.
        // Create useful helper objects for these.
        this.pathTemplates = {
            datasetPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/processors/{processor}/dataset'),
            datasetSchemaPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/processors/{processor}/dataset/datasetSchema'),
            evaluationPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/processors/{processor}/processorVersions/{processor_version}/evaluations/{evaluation}'),
            processorPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/processors/{processor}'),
            processorTypePathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/processorTypes/{processor_type}'),
            processorVersionPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/processors/{processor}/processorVersions/{processor_version}'),
        };
        // Some of the methods on this service return "paged" results,
        // (e.g. 50 results at a time, with tokens to get subsequent
        // pages). Denote the keys used for pagination and results.
        this.descriptors.page = {
            listDocuments: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'documentMetadata'),
        };
        const protoFilesRoot = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
        // This API contains "long-running operations", which return a
        // an Operation object that allows for tracking of the operation,
        // rather than holding a request open.
        const lroOptions = {
            auth: this.auth,
            grpc: 'grpc' in this._gaxGrpc ? this._gaxGrpc.grpc : undefined,
        };
        if (opts.fallback) {
            lroOptions.protoJson = protoFilesRoot;
            lroOptions.httpRules = [
                {
                    selector: 'google.cloud.location.Locations.GetLocation',
                    get: '/v1beta3/{name=projects/*/locations/*}',
                    additional_bindings: [
                        { get: '/uiv1beta3/{name=projects/*/locations/*}' },
                    ],
                },
                {
                    selector: 'google.cloud.location.Locations.ListLocations',
                    get: '/v1beta3/{name=projects/*}/locations',
                    additional_bindings: [
                        { get: '/uiv1beta3/{name=projects/*}/locations' },
                    ],
                },
                {
                    selector: 'google.longrunning.Operations.CancelOperation',
                    post: '/v1beta3/{name=projects/*/locations/*/operations/*}:cancel',
                    additional_bindings: [
                        {
                            post: '/uiv1beta3/{name=projects/*/locations/*/operations/*}:cancel',
                        },
                    ],
                },
                {
                    selector: 'google.longrunning.Operations.GetOperation',
                    get: '/v1beta3/{name=projects/*/locations/*/operations/*}',
                    additional_bindings: [
                        { get: '/uiv1beta3/{name=projects/*/locations/*/operations/*}' },
                    ],
                },
                {
                    selector: 'google.longrunning.Operations.ListOperations',
                    get: '/v1beta3/{name=projects/*/locations/*/operations}',
                    additional_bindings: [
                        { get: '/uiv1beta3/{name=projects/*/locations/*/operations}' },
                    ],
                },
            ];
        }
        this.operationsClient = this._gaxModule
            .lro(lroOptions)
            .operationsClient(opts);
        const updateDatasetResponse = protoFilesRoot.lookup('.google.cloud.documentai.v1beta3.Dataset');
        const updateDatasetMetadata = protoFilesRoot.lookup('.google.cloud.documentai.v1beta3.UpdateDatasetOperationMetadata');
        const importDocumentsResponse = protoFilesRoot.lookup('.google.cloud.documentai.v1beta3.ImportDocumentsResponse');
        const importDocumentsMetadata = protoFilesRoot.lookup('.google.cloud.documentai.v1beta3.ImportDocumentsMetadata');
        const batchDeleteDocumentsResponse = protoFilesRoot.lookup('.google.cloud.documentai.v1beta3.BatchDeleteDocumentsResponse');
        const batchDeleteDocumentsMetadata = protoFilesRoot.lookup('.google.cloud.documentai.v1beta3.BatchDeleteDocumentsMetadata');
        this.descriptors.longrunning = {
            updateDataset: new this._gaxModule.LongrunningDescriptor(this.operationsClient, updateDatasetResponse.decode.bind(updateDatasetResponse), updateDatasetMetadata.decode.bind(updateDatasetMetadata)),
            importDocuments: new this._gaxModule.LongrunningDescriptor(this.operationsClient, importDocumentsResponse.decode.bind(importDocumentsResponse), importDocumentsMetadata.decode.bind(importDocumentsMetadata)),
            batchDeleteDocuments: new this._gaxModule.LongrunningDescriptor(this.operationsClient, batchDeleteDocumentsResponse.decode.bind(batchDeleteDocumentsResponse), batchDeleteDocumentsMetadata.decode.bind(batchDeleteDocumentsMetadata)),
        };
        // Put together the default options sent with requests.
        this._defaults = this._gaxGrpc.constructSettings('google.cloud.documentai.v1beta3.DocumentService', gapicConfig, opts.clientConfig || {}, { 'x-goog-api-client': clientHeader.join(' ') });
        // Set up a dictionary of "inner API calls"; the core implementation
        // of calling the API is handled in `google-gax`, with this code
        // merely providing the destination and request information.
        this.innerApiCalls = {};
        // Add a warn function to the client constructor so it can be easily tested.
        this.warn = this._gaxModule.warn;
    }
    /**
     * Initialize the client.
     * Performs asynchronous operations (such as authentication) and prepares the client.
     * This function will be called automatically when any class method is called for the
     * first time, but if you need to initialize it before calling an actual method,
     * feel free to call initialize() directly.
     *
     * You can await on this method if you want to make sure the client is initialized.
     *
     * @returns {Promise} A promise that resolves to an authenticated service stub.
     */
    initialize() {
        // If the client stub promise is already initialized, return immediately.
        if (this.documentServiceStub) {
            return this.documentServiceStub;
        }
        // Put together the "service stub" for
        // google.cloud.documentai.v1beta3.DocumentService.
        this.documentServiceStub = this._gaxGrpc.createStub(this._opts.fallback
            ? this._protos.lookupService('google.cloud.documentai.v1beta3.DocumentService')
            : // eslint-disable-next-line @typescript-eslint/no-explicit-any
                this._protos.google.cloud.documentai.v1beta3.DocumentService, this._opts, this._providedCustomServicePath);
        // Iterate over each of the methods that the service provides
        // and create an API call method for each.
        const documentServiceStubMethods = [
            'updateDataset',
            'importDocuments',
            'getDocument',
            'listDocuments',
            'batchDeleteDocuments',
            'getDatasetSchema',
            'updateDatasetSchema',
        ];
        for (const methodName of documentServiceStubMethods) {
            const callPromise = this.documentServiceStub.then(stub => (...args) => {
                if (this._terminated) {
                    return Promise.reject('The client has already been closed.');
                }
                const func = stub[methodName];
                return func.apply(stub, args);
            }, (err) => () => {
                throw err;
            });
            const descriptor = this.descriptors.page[methodName] ||
                this.descriptors.longrunning[methodName] ||
                undefined;
            const apiCall = this._gaxModule.createApiCall(callPromise, this._defaults[methodName], descriptor, this._opts.fallback);
            this.innerApiCalls[methodName] = apiCall;
        }
        return this.documentServiceStub;
    }
    /**
     * The DNS address for this API service.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get servicePath() {
        if (typeof process === 'object' &&
            typeof process.emitWarning === 'function') {
            process.emitWarning('Static servicePath is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'documentai.googleapis.com';
    }
    /**
     * The DNS address for this API service - same as servicePath.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get apiEndpoint() {
        if (typeof process === 'object' &&
            typeof process.emitWarning === 'function') {
            process.emitWarning('Static apiEndpoint is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'documentai.googleapis.com';
    }
    /**
     * The DNS address for this API service.
     * @returns {string} The DNS address for this service.
     */
    get apiEndpoint() {
        return this._servicePath;
    }
    get universeDomain() {
        return this._universeDomain;
    }
    /**
     * The port for this API service.
     * @returns {number} The default port for this service.
     */
    static get port() {
        return 443;
    }
    /**
     * The scopes needed to make gRPC calls for every method defined
     * in this service.
     * @returns {string[]} List of default scopes.
     */
    static get scopes() {
        return ['https://www.googleapis.com/auth/cloud-platform'];
    }
    /**
     * Return the project ID used by this class.
     * @returns {Promise} A promise that resolves to string containing the project ID.
     */
    getProjectId(callback) {
        if (callback) {
            this.auth.getProjectId(callback);
            return;
        }
        return this.auth.getProjectId();
    }
    getDocument(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                dataset: (_a = request.dataset) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getDocument(request, options, callback);
    }
    getDatasetSchema(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getDatasetSchema(request, options, callback);
    }
    updateDatasetSchema(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'dataset_schema.name': (_a = request.datasetSchema.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateDatasetSchema(request, options, callback);
    }
    updateDataset(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'dataset.name': (_a = request.dataset.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateDataset(request, options, callback);
    }
    /**
     * Check the status of the long running operation returned by `updateDataset()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta3/document_service.update_dataset.js</caption>
     * region_tag:documentai_v1beta3_generated_DocumentService_UpdateDataset_async
     */
    async checkUpdateDatasetProgress(name) {
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.updateDataset, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    importDocuments(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                dataset: (_a = request.dataset) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.importDocuments(request, options, callback);
    }
    /**
     * Check the status of the long running operation returned by `importDocuments()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta3/document_service.import_documents.js</caption>
     * region_tag:documentai_v1beta3_generated_DocumentService_ImportDocuments_async
     */
    async checkImportDocumentsProgress(name) {
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.importDocuments, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    batchDeleteDocuments(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                dataset: (_a = request.dataset) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.batchDeleteDocuments(request, options, callback);
    }
    /**
     * Check the status of the long running operation returned by `batchDeleteDocuments()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta3/document_service.batch_delete_documents.js</caption>
     * region_tag:documentai_v1beta3_generated_DocumentService_BatchDeleteDocuments_async
     */
    async checkBatchDeleteDocumentsProgress(name) {
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.batchDeleteDocuments, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    listDocuments(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                dataset: (_a = request.dataset) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.listDocuments(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.dataset
     *   Required. The resource name of the dataset to be listed.
     *   Format:
     *   projects/{project}/locations/{location}/processors/{processor}/dataset
     * @param {number} request.pageSize
     *   The maximum number of documents to return. The service may return
     *   fewer than this value.
     *   If unspecified, at most 20 documents will be returned.
     *   The maximum value is 100; values above 100 will be coerced to 100.
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListDocuments` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListDocuments`
     *   must match the call that provided the page token.
     * @param {string} [request.filter]
     *   Optional. Query to filter the documents based on
     *   https://google.aip.dev/160.
     *   ## Currently support query strings are:
     *
     *   `SplitType=DATASET_SPLIT_TEST|DATASET_SPLIT_TRAIN|DATASET_SPLIT_UNASSIGNED`
     *   - `LabelingState=DOCUMENT_LABELED|DOCUMENT_UNLABELED|DOCUMENT_AUTO_LABELED`
     *   - `DisplayName=\"file_name.pdf\"`
     *   - `EntityType=abc/def`
     *   - `TagName=\"auto-labeling-running\"|\"sampled\"`
     *
     *   Note:
     *   - Only `AND`, `=` and `!=` are supported.
     *       e.g. `DisplayName=file_name AND EntityType!=abc` IS supported.
     *   - Wildcard `*` is supported only in `DisplayName` filter
     *   - No duplicate filter keys are allowed,
     *       e.g. `EntityType=a AND EntityType=b` is NOT supported.
     *   - String match is case sensitive (for filter `DisplayName` & `EntityType`).
     * @param {boolean} [request.returnTotalSize]
     *   Optional. Controls if the request requires a total size of matched
     *   documents. See
     *   {@link protos.google.cloud.documentai.v1beta3.ListDocumentsResponse.total_size|ListDocumentsResponse.total_size}.
     *
     *   Enabling this flag may adversely impact performance.
     *
     *   Defaults to false.
     * @param {number} [request.skip]
     *   Optional. Number of results to skip beginning from the `page_token` if
     *   provided. https://google.aip.dev/158#skipping-results. It must be a
     *   non-negative integer. Negative values will be rejected. Note that this is
     *   not the number of pages to skip. If this value causes the cursor to move
     *   past the end of results,
     *   {@link protos.google.cloud.documentai.v1beta3.ListDocumentsResponse.document_metadata|ListDocumentsResponse.document_metadata}
     *   and
     *   {@link protos.google.cloud.documentai.v1beta3.ListDocumentsResponse.next_page_token|ListDocumentsResponse.next_page_token}
     *   will be empty.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.cloud.documentai.v1beta3.DocumentMetadata|DocumentMetadata} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listDocumentsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listDocumentsStream(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                dataset: (_a = request.dataset) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listDocuments'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listDocuments.createStream(this.innerApiCalls.listDocuments, request, callSettings);
    }
    /**
     * Equivalent to `listDocuments`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.dataset
     *   Required. The resource name of the dataset to be listed.
     *   Format:
     *   projects/{project}/locations/{location}/processors/{processor}/dataset
     * @param {number} request.pageSize
     *   The maximum number of documents to return. The service may return
     *   fewer than this value.
     *   If unspecified, at most 20 documents will be returned.
     *   The maximum value is 100; values above 100 will be coerced to 100.
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListDocuments` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListDocuments`
     *   must match the call that provided the page token.
     * @param {string} [request.filter]
     *   Optional. Query to filter the documents based on
     *   https://google.aip.dev/160.
     *   ## Currently support query strings are:
     *
     *   `SplitType=DATASET_SPLIT_TEST|DATASET_SPLIT_TRAIN|DATASET_SPLIT_UNASSIGNED`
     *   - `LabelingState=DOCUMENT_LABELED|DOCUMENT_UNLABELED|DOCUMENT_AUTO_LABELED`
     *   - `DisplayName=\"file_name.pdf\"`
     *   - `EntityType=abc/def`
     *   - `TagName=\"auto-labeling-running\"|\"sampled\"`
     *
     *   Note:
     *   - Only `AND`, `=` and `!=` are supported.
     *       e.g. `DisplayName=file_name AND EntityType!=abc` IS supported.
     *   - Wildcard `*` is supported only in `DisplayName` filter
     *   - No duplicate filter keys are allowed,
     *       e.g. `EntityType=a AND EntityType=b` is NOT supported.
     *   - String match is case sensitive (for filter `DisplayName` & `EntityType`).
     * @param {boolean} [request.returnTotalSize]
     *   Optional. Controls if the request requires a total size of matched
     *   documents. See
     *   {@link protos.google.cloud.documentai.v1beta3.ListDocumentsResponse.total_size|ListDocumentsResponse.total_size}.
     *
     *   Enabling this flag may adversely impact performance.
     *
     *   Defaults to false.
     * @param {number} [request.skip]
     *   Optional. Number of results to skip beginning from the `page_token` if
     *   provided. https://google.aip.dev/158#skipping-results. It must be a
     *   non-negative integer. Negative values will be rejected. Note that this is
     *   not the number of pages to skip. If this value causes the cursor to move
     *   past the end of results,
     *   {@link protos.google.cloud.documentai.v1beta3.ListDocumentsResponse.document_metadata|ListDocumentsResponse.document_metadata}
     *   and
     *   {@link protos.google.cloud.documentai.v1beta3.ListDocumentsResponse.next_page_token|ListDocumentsResponse.next_page_token}
     *   will be empty.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.cloud.documentai.v1beta3.DocumentMetadata|DocumentMetadata}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta3/document_service.list_documents.js</caption>
     * region_tag:documentai_v1beta3_generated_DocumentService_ListDocuments_async
     */
    listDocumentsAsync(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                dataset: (_a = request.dataset) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listDocuments'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listDocuments.asyncIterate(this.innerApiCalls['listDocuments'], request, callSettings);
    }
    /**
     * Gets information about a location.
     *
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.name
     *   Resource name for the location.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html | CallOptions} for more details.
     * @returns {Promise} - The promise which resolves to an array.
     *   The first element of the array is an object representing {@link google.cloud.location.Location | Location}.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#regular-methods | documentation }
     *   for more details and examples.
     * @example
     * ```
     * const [response] = await client.getLocation(request);
     * ```
     */
    getLocation(request, options, callback) {
        return this.locationsClient.getLocation(request, options, callback);
    }
    /**
     * Lists information about the supported locations for this service. Returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.name
     *   The resource that owns the locations collection, if applicable.
     * @param {string} request.filter
     *   The standard list filter.
     * @param {number} request.pageSize
     *   The standard list page size.
     * @param {string} request.pageToken
     *   The standard list page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link google.cloud.location.Location | Location}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example
     * ```
     * const iterable = client.listLocationsAsync(request);
     * for await (const response of iterable) {
     *   // process response
     * }
     * ```
     */
    listLocationsAsync(request, options) {
        return this.locationsClient.listLocationsAsync(request, options);
    }
    /**
     * Gets the latest state of a long-running operation.  Clients can use this
     * method to poll the operation result at intervals as recommended by the API
     * service.
     *
     * @param {Object} request - The request object that will be sent.
     * @param {string} request.name - The name of the operation resource.
     * @param {Object=} options
     *   Optional parameters. You can override the default settings for this call,
     *   e.g, timeout, retries, paginations, etc. See {@link
     *   https://googleapis.github.io/gax-nodejs/global.html#CallOptions | gax.CallOptions}
     *   for the details.
     * @param {function(?Error, ?Object)=} callback
     *   The function which will be called with the result of the API call.
     *
     *   The second parameter to the callback is an object representing
     *   {@link google.longrunning.Operation | google.longrunning.Operation}.
     * @return {Promise} - The promise which resolves to an array.
     *   The first element of the array is an object representing
     * {@link google.longrunning.Operation | google.longrunning.Operation}.
     * The promise has a method named "cancel" which cancels the ongoing API call.
     *
     * @example
     * ```
     * const client = longrunning.operationsClient();
     * const name = '';
     * const [response] = await client.getOperation({name});
     * // doThingsWith(response)
     * ```
     */
    getOperation(request, options, callback) {
        return this.operationsClient.getOperation(request, options, callback);
    }
    /**
     * Lists operations that match the specified filter in the request. If the
     * server doesn't support this method, it returns `UNIMPLEMENTED`. Returns an iterable object.
     *
     * For-await-of syntax is used with the iterable to recursively get response element on-demand.
     *
     * @param {Object} request - The request object that will be sent.
     * @param {string} request.name - The name of the operation collection.
     * @param {string} request.filter - The standard list filter.
     * @param {number=} request.pageSize -
     *   The maximum number of resources contained in the underlying API
     *   response. If page streaming is performed per-resource, this
     *   parameter does not affect the return value. If page streaming is
     *   performed per-page, this determines the maximum number of
     *   resources in a page.
     * @param {Object=} options
     *   Optional parameters. You can override the default settings for this call,
     *   e.g, timeout, retries, paginations, etc. See {@link
     *   https://googleapis.github.io/gax-nodejs/global.html#CallOptions | gax.CallOptions} for the
     *   details.
     * @returns {Object}
     *   An iterable Object that conforms to {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | iteration protocols}.
     *
     * @example
     * ```
     * const client = longrunning.operationsClient();
     * for await (const response of client.listOperationsAsync(request));
     * // doThingsWith(response)
     * ```
     */
    listOperationsAsync(request, options) {
        return this.operationsClient.listOperationsAsync(request, options);
    }
    /**
     * Starts asynchronous cancellation on a long-running operation.  The server
     * makes a best effort to cancel the operation, but success is not
     * guaranteed.  If the server doesn't support this method, it returns
     * `google.rpc.Code.UNIMPLEMENTED`.  Clients can use
     * {@link Operations.GetOperation} or
     * other methods to check whether the cancellation succeeded or whether the
     * operation completed despite cancellation. On successful cancellation,
     * the operation is not deleted; instead, it becomes an operation with
     * an {@link Operation.error} value with a {@link google.rpc.Status.code} of
     * 1, corresponding to `Code.CANCELLED`.
     *
     * @param {Object} request - The request object that will be sent.
     * @param {string} request.name - The name of the operation resource to be cancelled.
     * @param {Object=} options
     *   Optional parameters. You can override the default settings for this call,
     * e.g, timeout, retries, paginations, etc. See {@link
     * https://googleapis.github.io/gax-nodejs/global.html#CallOptions | gax.CallOptions} for the
     * details.
     * @param {function(?Error)=} callback
     *   The function which will be called with the result of the API call.
     * @return {Promise} - The promise which resolves when API call finishes.
     *   The promise has a method named "cancel" which cancels the ongoing API
     * call.
     *
     * @example
     * ```
     * const client = longrunning.operationsClient();
     * await client.cancelOperation({name: ''});
     * ```
     */
    cancelOperation(request, options, callback) {
        return this.operationsClient.cancelOperation(request, options, callback);
    }
    /**
     * Deletes a long-running operation. This method indicates that the client is
     * no longer interested in the operation result. It does not cancel the
     * operation. If the server doesn't support this method, it returns
     * `google.rpc.Code.UNIMPLEMENTED`.
     *
     * @param {Object} request - The request object that will be sent.
     * @param {string} request.name - The name of the operation resource to be deleted.
     * @param {Object=} options
     *   Optional parameters. You can override the default settings for this call,
     * e.g, timeout, retries, paginations, etc. See {@link
     * https://googleapis.github.io/gax-nodejs/global.html#CallOptions | gax.CallOptions}
     * for the details.
     * @param {function(?Error)=} callback
     *   The function which will be called with the result of the API call.
     * @return {Promise} - The promise which resolves when API call finishes.
     *   The promise has a method named "cancel" which cancels the ongoing API
     * call.
     *
     * @example
     * ```
     * const client = longrunning.operationsClient();
     * await client.deleteOperation({name: ''});
     * ```
     */
    deleteOperation(request, options, callback) {
        return this.operationsClient.deleteOperation(request, options, callback);
    }
    // --------------------
    // -- Path templates --
    // --------------------
    /**
     * Return a fully-qualified dataset resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} processor
     * @returns {string} Resource name string.
     */
    datasetPath(project, location, processor) {
        return this.pathTemplates.datasetPathTemplate.render({
            project: project,
            location: location,
            processor: processor,
        });
    }
    /**
     * Parse the project from Dataset resource.
     *
     * @param {string} datasetName
     *   A fully-qualified path representing Dataset resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromDatasetName(datasetName) {
        return this.pathTemplates.datasetPathTemplate.match(datasetName).project;
    }
    /**
     * Parse the location from Dataset resource.
     *
     * @param {string} datasetName
     *   A fully-qualified path representing Dataset resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromDatasetName(datasetName) {
        return this.pathTemplates.datasetPathTemplate.match(datasetName).location;
    }
    /**
     * Parse the processor from Dataset resource.
     *
     * @param {string} datasetName
     *   A fully-qualified path representing Dataset resource.
     * @returns {string} A string representing the processor.
     */
    matchProcessorFromDatasetName(datasetName) {
        return this.pathTemplates.datasetPathTemplate.match(datasetName).processor;
    }
    /**
     * Return a fully-qualified datasetSchema resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} processor
     * @returns {string} Resource name string.
     */
    datasetSchemaPath(project, location, processor) {
        return this.pathTemplates.datasetSchemaPathTemplate.render({
            project: project,
            location: location,
            processor: processor,
        });
    }
    /**
     * Parse the project from DatasetSchema resource.
     *
     * @param {string} datasetSchemaName
     *   A fully-qualified path representing DatasetSchema resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromDatasetSchemaName(datasetSchemaName) {
        return this.pathTemplates.datasetSchemaPathTemplate.match(datasetSchemaName)
            .project;
    }
    /**
     * Parse the location from DatasetSchema resource.
     *
     * @param {string} datasetSchemaName
     *   A fully-qualified path representing DatasetSchema resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromDatasetSchemaName(datasetSchemaName) {
        return this.pathTemplates.datasetSchemaPathTemplate.match(datasetSchemaName)
            .location;
    }
    /**
     * Parse the processor from DatasetSchema resource.
     *
     * @param {string} datasetSchemaName
     *   A fully-qualified path representing DatasetSchema resource.
     * @returns {string} A string representing the processor.
     */
    matchProcessorFromDatasetSchemaName(datasetSchemaName) {
        return this.pathTemplates.datasetSchemaPathTemplate.match(datasetSchemaName)
            .processor;
    }
    /**
     * Return a fully-qualified evaluation resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} processor
     * @param {string} processor_version
     * @param {string} evaluation
     * @returns {string} Resource name string.
     */
    evaluationPath(project, location, processor, processorVersion, evaluation) {
        return this.pathTemplates.evaluationPathTemplate.render({
            project: project,
            location: location,
            processor: processor,
            processor_version: processorVersion,
            evaluation: evaluation,
        });
    }
    /**
     * Parse the project from Evaluation resource.
     *
     * @param {string} evaluationName
     *   A fully-qualified path representing Evaluation resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromEvaluationName(evaluationName) {
        return this.pathTemplates.evaluationPathTemplate.match(evaluationName)
            .project;
    }
    /**
     * Parse the location from Evaluation resource.
     *
     * @param {string} evaluationName
     *   A fully-qualified path representing Evaluation resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromEvaluationName(evaluationName) {
        return this.pathTemplates.evaluationPathTemplate.match(evaluationName)
            .location;
    }
    /**
     * Parse the processor from Evaluation resource.
     *
     * @param {string} evaluationName
     *   A fully-qualified path representing Evaluation resource.
     * @returns {string} A string representing the processor.
     */
    matchProcessorFromEvaluationName(evaluationName) {
        return this.pathTemplates.evaluationPathTemplate.match(evaluationName)
            .processor;
    }
    /**
     * Parse the processor_version from Evaluation resource.
     *
     * @param {string} evaluationName
     *   A fully-qualified path representing Evaluation resource.
     * @returns {string} A string representing the processor_version.
     */
    matchProcessorVersionFromEvaluationName(evaluationName) {
        return this.pathTemplates.evaluationPathTemplate.match(evaluationName)
            .processor_version;
    }
    /**
     * Parse the evaluation from Evaluation resource.
     *
     * @param {string} evaluationName
     *   A fully-qualified path representing Evaluation resource.
     * @returns {string} A string representing the evaluation.
     */
    matchEvaluationFromEvaluationName(evaluationName) {
        return this.pathTemplates.evaluationPathTemplate.match(evaluationName)
            .evaluation;
    }
    /**
     * Return a fully-qualified processor resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} processor
     * @returns {string} Resource name string.
     */
    processorPath(project, location, processor) {
        return this.pathTemplates.processorPathTemplate.render({
            project: project,
            location: location,
            processor: processor,
        });
    }
    /**
     * Parse the project from Processor resource.
     *
     * @param {string} processorName
     *   A fully-qualified path representing Processor resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromProcessorName(processorName) {
        return this.pathTemplates.processorPathTemplate.match(processorName)
            .project;
    }
    /**
     * Parse the location from Processor resource.
     *
     * @param {string} processorName
     *   A fully-qualified path representing Processor resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromProcessorName(processorName) {
        return this.pathTemplates.processorPathTemplate.match(processorName)
            .location;
    }
    /**
     * Parse the processor from Processor resource.
     *
     * @param {string} processorName
     *   A fully-qualified path representing Processor resource.
     * @returns {string} A string representing the processor.
     */
    matchProcessorFromProcessorName(processorName) {
        return this.pathTemplates.processorPathTemplate.match(processorName)
            .processor;
    }
    /**
     * Return a fully-qualified processorType resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} processor_type
     * @returns {string} Resource name string.
     */
    processorTypePath(project, location, processorType) {
        return this.pathTemplates.processorTypePathTemplate.render({
            project: project,
            location: location,
            processor_type: processorType,
        });
    }
    /**
     * Parse the project from ProcessorType resource.
     *
     * @param {string} processorTypeName
     *   A fully-qualified path representing ProcessorType resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromProcessorTypeName(processorTypeName) {
        return this.pathTemplates.processorTypePathTemplate.match(processorTypeName)
            .project;
    }
    /**
     * Parse the location from ProcessorType resource.
     *
     * @param {string} processorTypeName
     *   A fully-qualified path representing ProcessorType resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromProcessorTypeName(processorTypeName) {
        return this.pathTemplates.processorTypePathTemplate.match(processorTypeName)
            .location;
    }
    /**
     * Parse the processor_type from ProcessorType resource.
     *
     * @param {string} processorTypeName
     *   A fully-qualified path representing ProcessorType resource.
     * @returns {string} A string representing the processor_type.
     */
    matchProcessorTypeFromProcessorTypeName(processorTypeName) {
        return this.pathTemplates.processorTypePathTemplate.match(processorTypeName)
            .processor_type;
    }
    /**
     * Return a fully-qualified processorVersion resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} processor
     * @param {string} processor_version
     * @returns {string} Resource name string.
     */
    processorVersionPath(project, location, processor, processorVersion) {
        return this.pathTemplates.processorVersionPathTemplate.render({
            project: project,
            location: location,
            processor: processor,
            processor_version: processorVersion,
        });
    }
    /**
     * Parse the project from ProcessorVersion resource.
     *
     * @param {string} processorVersionName
     *   A fully-qualified path representing ProcessorVersion resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromProcessorVersionName(processorVersionName) {
        return this.pathTemplates.processorVersionPathTemplate.match(processorVersionName).project;
    }
    /**
     * Parse the location from ProcessorVersion resource.
     *
     * @param {string} processorVersionName
     *   A fully-qualified path representing ProcessorVersion resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromProcessorVersionName(processorVersionName) {
        return this.pathTemplates.processorVersionPathTemplate.match(processorVersionName).location;
    }
    /**
     * Parse the processor from ProcessorVersion resource.
     *
     * @param {string} processorVersionName
     *   A fully-qualified path representing ProcessorVersion resource.
     * @returns {string} A string representing the processor.
     */
    matchProcessorFromProcessorVersionName(processorVersionName) {
        return this.pathTemplates.processorVersionPathTemplate.match(processorVersionName).processor;
    }
    /**
     * Parse the processor_version from ProcessorVersion resource.
     *
     * @param {string} processorVersionName
     *   A fully-qualified path representing ProcessorVersion resource.
     * @returns {string} A string representing the processor_version.
     */
    matchProcessorVersionFromProcessorVersionName(processorVersionName) {
        return this.pathTemplates.processorVersionPathTemplate.match(processorVersionName).processor_version;
    }
    /**
     * Terminate the gRPC channel and close the client.
     *
     * The client will no longer be usable and all future behavior is undefined.
     * @returns {Promise} A promise that resolves when the client is closed.
     */
    close() {
        if (this.documentServiceStub && !this._terminated) {
            return this.documentServiceStub.then(stub => {
                this._terminated = true;
                stub.close();
                this.locationsClient.close();
                this.operationsClient.close();
            });
        }
        return Promise.resolve();
    }
}
exports.DocumentServiceClient = DocumentServiceClient;
